xclim/sdba/adjustment.py

```diff
@@ -506,3 +510,133 @@ def _train(self, ref, hist):
+      which case all dimensions except "time" are used. The dask version of the
+      training algorithm currently doesn't support any chunking along the coordinates
+      and the time dimensions.
+
+
+    In adjustment:
+
+    norm_to : {'hist', 'sim'}
+      Before the transformation, sim values are normalized by subtracting the mean of
+      hist if norm_to == 'hist' (default) and its own mean if norm_to == 'sim'.
+
+    Notes
+    -----
+    The input data is transformed of N points in a :math:`M`-dimensional space.
+    Where :math:`N` is taken along the 'time' coordinates, but :math:`M` can be the
+    concatenation of any number of pre-existing dimensions (the default being all
+    except 'time'). Thus, the adjustment is equivalent to a linear transformation
+    of these :math:`N` points in a the :math:`M`-dimensional space.
+
+    The principal components (PC) of `hist` and `ref` are used to defined new
+    coordinate systems, centered on their respective means. The training step creates a
+    matrix defining the transformation from `hist` to `ref`:
+
+    .. math::
+
+      scen = e_{R} + \mathrm{\mathbf{T}}(sim - e_{H})
+
+    Where:
+
+    .. math::
+
+      \mathrm{\mathbf{T}} = \mathrm{\mathbf{R}}\mathrm{\mathbf{H}}^{-1}
+
+    :math:`\mathrm{\mathbf{R}}` is the matrix transforming from the PC coordinates
+    computed on `ref` to the data coordinates. Similarly, :math:`\mathrm{\mathbf{H}}`
+    is transform from the `hist` PC to the data coordinates
+    (:math:`\mathrm{\mathbf{H}}` is the inverse transformation). :math:`e_R` and
+    :math:`e_H` are the centroids of the `ref` and `hist` distributions respectively.
+    Upon running the  `adjust` step, one may decide to use :math:`e_S`, the centroid
+    of the `sim` distribution, instead of :math:`e_H`.
+
+    References
+    ----------
+    .. [hnilica2017] Hnilica, J., Hanel, M. and Puš, V. (2017), Multisite bias correction of precipitation data from regional climate models. Int. J. Climatol., 37: 2934-2946. https://doi.org/10.1002/joc.4890
+    """
+
+    def __init__(self, *, dims=None):
+        super().__init__(dims=dims)
+
+    @staticmethod
+    def _manage_chunking(ds, dim):
+        if isinstance(ds.data, dsk.Array):
+            if len(ds.chunks[ds.get_axis_num(dim)]) > 1:
+                warn("Merging chunks along coordinate dimensions for PC computation.")
+                return ds.chunk({dim: -1})
+        return ds
+
+    def _train(self, ref, hist):
+        dims = self.dims or (set(ref.dims) - {"time"})
+
+        # The multiple PC-space dimensions are along "coordinate"
+        # Matrix multiplication in xarray behaves as a dot product across
+        # same-name dimensions, instead of reducing according to the dimension order,
+        # as in numpy or normal maths. So crdX all refer to the same dimension,
+        # but with names assuring correct matrix multiplication even if they are out of order.
+        ref = ref.stack(crd1=dims)
```

Comment:
This will fail if `crd1` or `crd3` are names of existing dimensions. Unlikely, I know, but we could take the habit of using UUIDs or some other convention (underscore) to indicate temporary *work* dimensions.

Reply:
Didn't think of that! I'll use temp dim names.